src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs

```diff
@@ -43,34 +43,35 @@ internal TestExecutionManager(IEnvironment environment, Func<Action, Task>? task
         TestMethodFilter = new TestMethodFilter();
         _sessionParameters = new Dictionary<string, object>();
         _environment = environment;
-        _taskFactory = taskFactory ?? ThreadingHelper.RunAsync;
+        _taskFactory = taskFactory ?? DefaultFactoryAsync;
     }
 
-    internal static class ThreadingHelper
+    private static Task DefaultFactoryAsync(Action action)
     {
-        public static Task RunAsync(Action action)
+        if (MSTestSettings.RunConfigurationSettings.ExecutionApartmentState == ApartmentState.STA
+            && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
         {
-            if (MSTestSettings.RunConfigurationSettings.ExecutionApartmentState == ApartmentState.STA
-                && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            TaskCompletionSource<int> tcs = new();
```

Comment:
Is this deliberately not using the asynchronous continuation option? If so would it be worthwhile to add a comment of why?

https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#always-create-taskcompletionsourcet-with-taskcreationoptionsruncontinuationsasynchronously

Reply:
We don't have any continuation on this task but a wait all when the queue is empty https://github.com/microsoft/testfx/blob/main/src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs#L364

Reply:
Some more thoughts on it, MSTest code is pretty old and we're trying to move it to be fully async, we're not yet there so despite you can define a `Task` return test is handled in a suboptimal way.

cc: @Evangelink